File: server/routes/watchlists.js
import { Router } from 'express';
import prisma from '../lib/prisma.js';
import { requireAuth } from '../middleware/auth.js';

const router = Router();
router.use(requireAuth);

router.get('/', async (req, res, next) => {
  try {
    const watchlists = await prisma.watchlist.findMany({
      where: { userId: req.userId },
      include: { items: true },
      orderBy: { createdAt: 'desc' },
    });
    res.json(watchlists);
  } catch (err) { next(err); }
});

router.post('/', async (req, res, next) => {
  try {
    const { name } = req.body;
    if (!name) return res.status(400).json({ error: 'Name required' });
    const watchlist = await prisma.watchlist.create({
      data: { userId: req.userId, name },
      include: { items: true },
    });
    res.status(201).json(watchlist);
  } catch (err) { next(err); }
});

router.post('/:id/items', async (req, res, next) => {
  try {
    const { symbol, assetType } = req.body;
    if (!symbol) return res.status(400).json({ error: 'Symbol required' });
    const item = await prisma.watchlistItem.create({
      data: { watchlistId: req.params.id, symbol: symbol.toUpperCase(), assetType: assetType || 'stock' },
    });
    res.status(201).json(item);
  } catch (err) { next(err); }
});

router.delete('/:id', async (req, res, next) => {
  try {
    await prisma.watchlist.delete({ where: { id: req.params.id, userId: req.userId } });
    res.status(204).end();
  } catch (err) { next(err); }
});

router.delete('/:watchlistId/items/:itemId', async (req, res, next) => {
  try {
    // Verify item belongs to user's watchlist before deleting
    const item = await prisma.watchlistItem.findFirst({
      where: { id: req.params.itemId, watchlist: { userId: req.userId } },
    });
    if (!item) return res.status(404).json({ error: 'Item not found' });
    await prisma.watchlistItem.delete({ where: { id: req.params.itemId } });
    res.status(204).end();
  } catch (err) { next(err); }
});

export default router;

File: backend/app/routes/watchlists.py
import logging

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.database import get_db
from app.middleware.auth import AuthUser, require_auth
from app.models.watchlist import Watchlist, WatchlistItem
from app.schemas.watchlist import (
    WatchlistCreate,
    WatchlistItemCreate,
    WatchlistItemResponse,
    WatchlistResponse,
)
from app.services.audit import write_audit_log

logger = logging.getLogger(__name__)
router = APIRouter()


@router.get("/watchlists", response_model=list[WatchlistResponse])
async def list_watchlists(
    auth: AuthUser = Depends(require_auth),
    session: AsyncSession = Depends(get_db),
    limit: int = Query(50, ge=1, le=200, description="Max items to return"),
    offset: int = Query(0, ge=0, description="Number of items to skip"),
):
    result = await session.execute(
        select(Watchlist)
        .where(Watchlist.user_id == auth.user_id)
        .options(selectinload(Watchlist.items))
        .order_by(Watchlist.created_at.desc())
        .limit(limit)
        .offset(offset)
    )
    return list(result.scalars().all())


@router.post("/watchlists", response_model=WatchlistResponse, status_code=201)
async def create_watchlist(
    body: WatchlistCreate,
    auth: AuthUser = Depends(require_auth),
    session: AsyncSession = Depends(get_db),
):
    watchlist = Watchlist(user_id=auth.user_id, name=body.name)
    session.add(watchlist)
    await session.flush()
    await session.refresh(watchlist, ["items"])
    await write_audit_log(session, auth.user_id, "create", "watchlist", {"id": watchlist.id})
    return watchlist


@router.post("/watchlists/{watchlist_id}/items", response_model=WatchlistItemResponse, status_code=201)
async def add_watchlist_item(
    watchlist_id: str,
    body: WatchlistItemCreate,
    auth: AuthUser = Depends(require_auth),
    session: AsyncSession = Depends(get_db),
):
    # Verify ownership
    result = await session.execute(
        select(Watchlist).where(Watchlist.id == watchlist_id, Watchlist.user_id == auth.user_id)
    )
    watchlist = result.scalar_one_or_none()
    if not watchlist:
        raise HTTPException(status_code=404, detail="Resource not found")

    item = WatchlistItem(
        watchlist_id=watchlist_id,
        symbol=body.symbol.upper(),
        asset_type=body.asset_type,
    )
    session.add(item)
    await session.flush()
    return item


@router.delete("/watchlists/{watchlist_id}", status_code=204)
async def delete_watchlist(
    watchlist_id: str,
    auth: AuthUser = Depends(require_auth),
    session: AsyncSession = Depends(get_db),
):
    result = await session.execute(
        select(Watchlist).where(Watchlist.id == watchlist_id, Watchlist.user_id == auth.user_id)
    )
    watchlist = result.scalar_one_or_none()
    if not watchlist:
        raise HTTPException(status_code=404, detail="Resource not found")

    await session.delete(watchlist)
    await write_audit_log(session, auth.user_id, "delete", "watchlist", {"id": watchlist_id})


@router.delete("/watchlists/{watchlist_id}/items/{item_id}", status_code=204)
async def remove_watchlist_item(
    watchlist_id: str,
    item_id: str,
    auth: AuthUser = Depends(require_auth),
    session: AsyncSession = Depends(get_db),
):
    # Verify ownership
    result = await session.execute(
        select(Watchlist).where(Watchlist.id == watchlist_id, Watchlist.user_id == auth.user_id)
    )
    watchlist = result.scalar_one_or_none()
    if not watchlist:
        raise HTTPException(status_code=404, detail="Resource not found")

    result = await session.execute(
        select(WatchlistItem).where(
            WatchlistItem.id == item_id, WatchlistItem.watchlist_id == watchlist_id
        )
    )
    item = result.scalar_one_or_none()
    if not item:
        raise HTTPException(status_code=404, detail="Resource not found")

    await session.delete(item)
